### ISSUE TITLE
instrument window's width is too small

### ISSUE THREAD

**Issue by StCyr** [LMMS/lmms#2528]
Hello,

I'm creating this issue in order to keep track of the discussions held about this problem.

I originally reported this problem in issue in https://github.com/LMMS/lmms/issues/2519. In that issue, it as been considered as a translation issue and has been solved accordingly.

Though, not everybody agree about the real cause of this problem (me being the first tbh).

Here are the opinions that were exchanged on the IRC about this problem so far:

(2016/01/27)
**15:48:03 - StCyr: bah I'll send you an updated fr translation**
15:48:27 - StCyr: but why aren't these windows resizeable?
15:49:16 - StCyr: I understand it wouldn't probably be very nice when resized, but hey what a nightmare to see this tab
**15:49:51 - umcaruje: its a bad idea to make the window resizeable**
15:50:12 - umcaruje: its up to the translator to use abbrevation to keep the width small
15:51:00 - umcaruje: bad design, maybe, but it most probably won't change
15:51:26 - StCyr: when you say "bad idea" you mean "would become ugly"?
15:51:37 - StCyr: (that's not such a bad idea)
15:51:50 - StCyr: (well at least I can understand it)
15:54:20 - umcaruje: well, what would happen to the instrument GUI?
**15:54:36 - umcaruje: we'd need to add padding, and it would get really really ugly
15:55:21 - umcaruje: so yeah, if we actually had 2 rows
15:55:48 - umcaruje: with 3 items of the same, constant width, that'd be better tbh
15:57:51 - StCyr: that maybe an easier (and perfectly valid) fix
15:59:45 - StCyr: well not easier than updating the translation, but more robust
15:59:50 - umcaruje: yeah**
...
(2016/01/28)
16:38:26 - StCyr1: should some developer wanna have a look: https://github.com/StCyr/lmms/tree/StCyr_2ndRow
16:38:43 - StCyr1: (for a better solution to https://github.com/LMMS/lmms/issues/2519)
**16:51:03 - grejppi: StCyr1: tbh adding a second row feels like only treating a symptom and not the cause which is the tiny size of instrument windows**
16:52:06 - grejppi: the piano roll toolbar was recently split into two rows and it looks absolutely hideous
16:54:55 - umcaruje: grejppi: I'm the cause of that lol
**16:55:18 - umcaruje: I said that instruments windows shouldn't be resizeable**
**16:55:32 - umcaruje: since of the fixed size that the artwork has**
16:55:38 - grejppi: well it wouldn't make sense because of that
16:55:48 - umcaruje: yeah
**16:55:58 - umcaruje: so I said, if we actually had 2 rows
16:56:11 - umcaruje: where we had 3 buttons of the same width
16:56:16 - umcaruje: that could fix the problem**
**16:56:49 - grejppi: tabs on two rows is never a good thing**
**16:57:29 - umcaruje: but its not a problem in french only**
**17:14:55 - umcaruje: so StCyr1 I built your branch
17:15:57 - umcaruje: I don't really think this is a solution
17:16:06 - umcaruje: after seeing the extra space it adds**
17:16:23 - grejppi: shrink the text size then? GENIUS
17:18:21 - grejppi: using abbreviations would probably the most sane thing to do
17:30:17 - Babbage_: icons
17:34:27 - Babbage_: plug icon, envelope icon with a wave for on int, `Fn`, `FX` both common in the world, 5 pin mini plug, and something

**Comment by claell** [comment]
How does other software handle this? Are the windows resizeable there? Options would be to scale some parts, to rearrange some icons etc.

**Comment by StCyr** [comment]
Although, I don't want to continue arguing about this point, and I totally accept and respect grejpii and umcaruje's opinions, I would like to add some information to the discussion:

1. Although my solution currently adds 20 pixels to the instrument views' height, the defintive solution should only add max. 11 pixels per row, and should ideally only create additionnal row(s) when needed (so, english locale wouldn't be impacted);
2. We could also imagine other solutions like having arrows to scroll the tab bar, or make use of Qt's QTabBar widget which has a "edible" option so that tabs automatically shrink when there's not enough space to display them in their entirety 
3. I'm also not in favour of full translation (eg: "FX" should stay "FX", "BW" is perfectly understable by music savy french people, so is "pitch" and so on...)
4. Regarding, the piano roll toolbar that now looks hideous to grejppi: Maybe the piano roll could automatically scroll so that the played note always gets displayed? But, first, how was it before? I mean what was the solution before to display such a large keyboard? Also, what's the purpose of this toolbar anyway? Do people really use it to play notes? If yes, well, I understand automatic scrolling isn't the ultimate answer.

**Comment by claell** [comment]
Could it be an option to simply automatically reduce/increase padding adaptively to the length of words, so they exactly fit the space given? Also the font size could be lowered (adaptively to the needed size).

**Comment by StCyr** [comment]
To reply to claell: I have no idea how other softwares do that. 

My guess is that they also try to manage that with translation or icons as much as they can.

But, also, that they use several rows when their translations become so short that they don't make sense anymore

**Comment by StCyr** [comment]
I think that padding is actually automatically adapted to the length of words (I'm not sure what you mean here): Tabs' width currently depends on tab's label (ie: a tab with a shorter label, like "FX", is shorter than a tab with a longer label like "ENV/LFO")

**Comment by StCyr** [comment]
For reference, should I delete my branch, here's what I did so far regarding the automatic addition of rows:

```
--- a/src/gui/widgets/TabWidget.cpp
+++ b/src/gui/widgets/TabWidget.cpp
@@ -72,8 +72,8 @@ void TabWidget::addTab( QWidget * _w, const QString & _name, int _idx )
                }
        }
        m_widgets[_idx] = d;
-       _w->setFixedSize( width() - 4, height() - 14 );
-       _w->move( 2, 13 );
+       _w->setFixedSize( width() - 4, height() - 34 );
+       _w->move( 2, 33 );
        _w->hide();

        if( m_widgets.contains( m_activeTab ) )
@@ -108,20 +108,37 @@ void TabWidget::setActiveTab( int _idx )

 void TabWidget::mousePressEvent( QMouseEvent * _me )
 {
-       if( _me->y() > 1 && _me->y() < 13 )
+       if( _me->y() > 1 && _me->y() < 13 + m_tabheight )
        {
+                int twidth = 0;
+                int row = 1;
                int cx = ( ( m_caption == "" ) ? 4 : 14 ) +
                                        fontMetrics().width( m_caption );
+
+                // Parse the tab widgets to find out which one was clicked on
                for( widgetStack::iterator it = m_widgets.begin();
                                                it != m_widgets.end(); ++it )
                {
+                        // Switch to next row when we reach maximum width
+                        twidth += ( *it ).nwidth;
+                        if ( twidth >= width() )
+                        {
+                                cx = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
+                                row++;
+                                twidth = 0;
+                        }
+
                        if( _me->x() >= cx &&
-                                       _me->x() <= cx + ( *it ).nwidth )
+                                       _me->x() <= cx + ( *it ).nwidth &&
+                                        _me->y() >= 1 + m_tabheight * ( row - 1 ) &&
+                                        _me->y() <= 1 + m_tabheight * row )
+
                        {
                                setActiveTab( it.key() );
                                update();
                                return;
                        }
+
                        cx += ( *it ).nwidth;
                }
        }
@@ -169,7 +186,7 @@ void TabWidget::paintEvent( QPaintEvent * _pe )
        p.drawRect( 1, 1, width() - 3, height() - 3 );

        p.fillRect( 2, 2, width() - 4, m_tabheight, g );
-       p.drawLine( 2, m_tabheight + 2, width() - 3, m_tabheight + 2);
+       p.drawLine( 2, m_tabheight * 2 + 2, width() - 3, m_tabheight * 2 + 2);

        if( ! m_caption.isEmpty() )
        {
@@ -192,17 +209,31 @@ void TabWidget::paintEvent( QPaintEvent * _pe )
        }
        p.setPen( cap_col );

-
+        int twidth = 0;
+        int row = 1;
        for( widgetStack::iterator it = m_widgets.begin();
                                                it != m_widgets.end(); ++it )
        {
+                // Create a new row of tabs when there is no more space
+                twidth += ( *it ).nwidth;
+                if ( twidth >= width() )
+                {
+                        tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
+                        row++;
+                        twidth = 0;
+                }
+
+                // highlight the active tab
                if( it.key() == m_activeTab )
                {
                        p.setPen( QColor( 32, 48, 64 ) );
-                       p.fillRect( tab_x_offset, 2, ( *it ).nwidth - 6, 10, cap_col );
+                       p.fillRect( tab_x_offset, 2 + m_tabheight * ( row - 1 ), ( *it ).nwidth - 6, 10, cap_col );
                }
-               p.drawText( tab_x_offset + 3, m_tabheight, ( *it ).name );
+
+                // draw tab's label
+               p.drawText( tab_x_offset + 3, m_tabheight * row , ( *it ).name );
                p.setPen( cap_col );
+
                tab_x_offset += ( *it ).nwidth;
        }
 }
diff --git a/src/tracks/InstrumentTrack.cpp b/src/tracks/InstrumentTrack.cpp
index 545c82c..96fe01c 100644
--- a/src/tracks/InstrumentTrack.cpp
+++ b/src/tracks/InstrumentTrack.cpp
@@ -1407,7 +1407,7 @@ InstrumentTrackWindow::InstrumentTrackWindow( InstrumentTrackView * _itv ) :


        m_tabWidget = new TabWidget( "", this );
-       m_tabWidget->setFixedHeight( INSTRUMENT_HEIGHT + 10 );
+       m_tabWidget->setFixedHeight( INSTRUMENT_HEIGHT + 30 );


        // create tab-widgets
(END)
+       p.drawLine( 2, m_tabheight * 2 + 2, width() - 3, m_tabheight * 2 + 2);

        if( ! m_caption.isEmpty() )
        {
@@ -192,17 +209,31 @@ void TabWidget::paintEvent( QPaintEvent * _pe )
        }
        p.setPen( cap_col );

-
+        int twidth = 0;
+        int row = 1;
        for( widgetStack::iterator it = m_widgets.begin();
                                                it != m_widgets.end(); ++it )
        {
+                // Create a new row of tabs when there is no more space
+                twidth += ( *it ).nwidth;
+                if ( twidth >= width() )
+                {
+                        tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
+                        row++;
+                        twidth = 0;
+                }
+
+                // highlight the active tab
                if( it.key() == m_activeTab )
                {
                        p.setPen( QColor( 32, 48, 64 ) );
-                       p.fillRect( tab_x_offset, 2, ( *it ).nwidth - 6, 10, cap_col );
+                       p.fillRect( tab_x_offset, 2 + m_tabheight * ( row - 1 ), ( *it ).nwidth - 6, 10, cap_col );
                }
-               p.drawText( tab_x_offset + 3, m_tabheight, ( *it ).name );
+
+                // draw tab's label
+               p.drawText( tab_x_offset + 3, m_tabheight * row , ( *it ).name );
                p.setPen( cap_col );
+
                tab_x_offset += ( *it ).nwidth;
        }
 }
diff --git a/src/tracks/InstrumentTrack.cpp b/src/tracks/InstrumentTrack.cpp
index 545c82c..96fe01c 100644
--- a/src/tracks/InstrumentTrack.cpp
+++ b/src/tracks/InstrumentTrack.cpp
@@ -1407,7 +1407,7 @@ InstrumentTrackWindow::InstrumentTrackWindow( InstrumentTrackView * _itv ) :

        m_tabWidget = new TabWidget( "", this );
-       m_tabWidget->setFixedHeight( INSTRUMENT_HEIGHT + 10 );
+       m_tabWidget->setFixedHeight( INSTRUMENT_HEIGHT + 30 );


        // create tab-widgets

```

**Comment by claell** [comment]
With padding I meant the space between the words. It an still get a bit smaller. Also one could change it so that the whole space of the line gets used, even if the words are smaller. Currently in the English version there is some space on the right.

**Comment by BaraMGB** [comment]
Resizing the instrument window is not possible at the moment. The most instruments use move() for layout their knobs and buttons. And the background is a bitmap. Perhaps we can double the width and show at the right side the envelope tab. Gradually we redesign the instruments to double width.

**Comment by Umcaruje** [comment]
>Perhaps we can double the width and show at the right side the envelope tab. 

Yeah, that's another idea that comes to mind. If done well, it could be a winner.

**Comment by grejppi** [comment]
> Regarding, the piano roll toolbar that now looks hideous to grejppi: Maybe the piano roll could automatically scroll so that the played note always gets displayed? But, first, how was it before? I mean what was the solution before to display such a large keyboard? Also, what's the purpose of this toolbar anyway? Do people really use it to play notes? If yes, well, I understand automatic scrolling isn't the ultimate answer.

I was referring to this one:

![kuvakaappaus - 2016-01-29 17-39-22](https://cloud.githubusercontent.com/assets/772336/12679882/45ce0ffc-c6af-11e5-8c86-a434f3d36d1e.png)

**Comment by tresf** [comment]
Overflowing tabs [isn't a new concept](https://bugzilla.mozilla.org/show_bug.cgi?id=570564).  In regards to avoiding the translation, perhaps we entertain wordless artwork in its place and remove the text altogether.

**Comment by Spekular** [comment]
@tresf that sounds like a good idea

**Comment by StCyr** [comment]
Hello,

I must say I'm not convinced at all by this idea of artwork. 

First, I don't think wordless artworks would make this tab bar's height smaller than 2 rows of icons.
Secondly, I don't think it will be that easy to find meaningfull artworks for all the tabs.

Though, I would like to continue this discussion in the hope to find a solution to this overflowing tab bar. 

So, here are some possible artwork ideas that I found on the Internet:

- PLUGIN: Often represented as a piece of a puzzle, or a power plug. Maybe in this case it could be a house (meaning "home of the instrument window" ) 
- ENV/LFO: Should be something like a sine or sawtooth wave
- FX: I haven't found any drawing that directly make me think about "effects". But, maybe a simple artwork around the "FX" letters should be OK in any language.
- MIDI: Usually represented as a keyboard, or by a MIDI (DIN 41524) connector
- MISC: A star or a plus sign?

The other solutions evoked are:

1. Doubling the intrument window's width (BaraMGB, Umcaruje);
2. Using QTabBar widgets and use their "edible" property (StCyr);
3. Change padding and font size (claell);
4. Do not translate (StCyr);
5. A second row of icons (StCyr).

Cheers,

Cyril

**Comment by Spekular** [comment]
Maybe I'm wrong here but I don't think the icons need to exactly convey the
purpose of the tab. They should have some connection, but if were moving
away from text than we're moving from the only medium that can really
convey it. For that reason I think perhaps they should be more abstract,
for example:

*  Plugin: To me this is the generator, which is where we get the base
signal. Therefore I'd use a circle or circle with turbine.

*  Env/Lfo: This is where the sound is shaped, so an envelope (not the kind
you put letter in!) would make sense

*  FX: This one is tough, but an X would probably work

*  Midi: Keyboard or square

*  Misc: A plus

The only thing I'm not sure of is how to make these look good together, brb
with mockup.

**Comment by Spekular** [comment]
Here are small/abstract icons:
![new icons v1](https://cloud.githubusercontent.com/assets/8918426/12841345/7785bcfc-cbeb-11e5-8391-81c1bdb36587.png)

They don't look incredible but I'd say that's down to my mockup skills (no antialiasing etc.) and not the concept. They could probably actually be drawn with a font/vectors which is an added bonus.

**Comment by StCyr** [comment]
I like your idea @Spekular 

just simple symbols

... connected to (translated?) tooltips 

... with the currently selected symbol highlighted

...

maybe the artwork fans will win this argument :-)

....

Hopefully their will be winners, and a solution will be implemented :-)

Cheers,

Cyrille

**Comment by tresf** [comment]
> Here are small/abstract icons:

If we go this direction, I would advise we make the tabs slightly taller to accommodate a legible icon.  We can still use the translation for a tooltip.  Here are some placeholders.  tagging @RebeccaDeField.

* **Plugin Tab:** Puzzle piece
* **ENV/LFO:** Wave (as @Spekular illustrated)
* **FUNC:** Gear
* **FX:** Mixer icon
* **MIDI:** Plug or round midi icon

**Comment by RebeccaDeField** [comment]
Hopefully, producers should be able to get an idea of what each icon is supposed to mean at a glance, so somthing literal and simple is always a good idea. I'm not a producer myself, so I can't judge what icons would make the most sense for this instance, but I'm sure you guys can work that out.

We will probably want a little more space on that bar for legibility.

I agree that using a tooltip is also a good idea.

**Comment by StCyr** [comment]
Hi guys,

I just wanted to let you know that I've started a possible implementation of "artworkable" TabWidgets.

I've a working PoC here: https://github.com/StCyr/lmms/tree/StCyr_ArtworkTabs

Please tell me what you think about it.

Regards,

Cyrille

**Comment by StCyr** [comment]
Hi,

I've updated my artwork tab PoC so that it can handle both pixmap and text tabs.

It works as follow:

1* When given a pixmapName, TabWidget will draw an artwork tab using "pixmapName"
2* When pixmapName is NULL, Tabwidget will draw a text tab using "_name"

PoC is here: https://github.com/StCyr/lmms/tree/StCyr_ArtworkTabs

It seems that I failed to properly compile my previous commit. It will compile properly now.

Please tell me what you think about it.

Regards,

Cyrille

**Comment by Spekular** [comment]
That sounds like a pretty grear way to implement it!

**Comment by StCyr** [comment]
Cool,

The problem now is that I'm really not a design guy. So, creating new artworks might be terrible. But, I can try. 

Anyway, I need new artworks to find out how to embed them in LMMS binary. So, I'll create (probably awful) new ones.

Any idea for a good icon's size? 

Currently, the TabWidget for the "PLUGIN", "ENV/LFO", "FUNC",... tabs is 14 pixels tall, and 254 pixels large if I'm not mistaken. And, there are 6 tabs to be displayed.

Cyrille

**Comment by StCyr** [comment]
Ok,

so I've given a try with 14x28 icons. Here there are:

![midi_active](https://cloud.githubusercontent.com/assets/8179031/13148824/4224162a-d65f-11e5-97fd-7370b99cef06.png)
![midi_inactive](https://cloud.githubusercontent.com/assets/8179031/13148828/4229559a-d65f-11e5-9a6b-dca58add2a2e.png)
![plugin_active](https://cloud.githubusercontent.com/assets/8179031/13148829/42299302-d65f-11e5-9836-cd18b68646be.png)
![plugin_inactive](https://cloud.githubusercontent.com/assets/8179031/13148827/42279ad4-d65f-11e5-8bab-8273f2c740d8.png)
![fx_active](https://cloud.githubusercontent.com/assets/8179031/13148826/42276b4a-d65f-11e5-8b07-aeca680fe845.png)
![fx_inactive](https://cloud.githubusercontent.com/assets/8179031/13148825/422705ec-d65f-11e5-86ba-a22e3ad62245.png)
![env_lfo_active](https://cloud.githubusercontent.com/assets/8179031/13148830/42362a22-d65f-11e5-8c5b-28c291efa1cd.png)
![env_lfo_inactive](https://cloud.githubusercontent.com/assets/8179031/13148834/423a5048-d65f-11e5-9f47-761e9a6c850b.png)
![miscellaneous_active](https://cloud.githubusercontent.com/assets/8179031/13148831/42370c08-d65f-11e5-8599-149466d10966.png)
![miscellaneous_inactive](https://cloud.githubusercontent.com/assets/8179031/13148833/423990d6-d65f-11e5-8058-eabaa105f522.png)
![functions_active](https://cloud.githubusercontent.com/assets/8179031/13148832/4238dce0-d65f-11e5-84a0-a3668430bf9b.png)
![functions_inactive](https://cloud.githubusercontent.com/assets/8179031/13148835/423b9890-d65f-11e5-888f-55a8dbdd9584.png)

PoC has been updated: https://github.com/StCyr/lmms/tree/StCyr_ArtworkTabs

PS: I told you they would be ugly ;-)

**Comment by claell** [comment]
But they show the concept (and I don't think they are ugly). Maybe if this should get integrated into the main version, some designers will create some?

**Comment by tresf** [comment]
These are great.  Mockup?

**Comment by RebeccaDeField** [comment]
@StCyr Actually, that is a great start. Maybe you have more designer in your than you thought ;)

I wouldn't mind creating vector versions when I have a few minutes to spare, but I first need to know if the icons he attached make sense to everyone.

**Comment by StCyr** [comment]
Hi, 

Thanks for the compliment :-)

@tresf : I'm not sure I understand what you mean (french native here). If this is the source images that you want, I've created them with gimp and can attach them here tomorrow CEST. But, too bad, I didn't bother using 2 layers to ease rework  :-/

I'll also try to add tooltips; I doesn't look too complicated (http://doc.qt.io/qt-4.8/qt-widgets-tooltips-example.html)

Cyrille

**Comment by tresf** [comment]
> @tresf : I'm not sure I understand what you mean

Sorry, I meant a sample of these new icons in action (on the actual tabs on a sample instrument) :+1:

**Comment by StCyr** [comment]
Ah, ok.

Sure, I can take a snapshot tomorrow also.

Otherwise my branch https://github.com/StCyr/lmms/tree/StCyr_ArtworkTabs compiles and shows the artwork tabs in action :-)

C.

**Comment by StCyr** [comment]
Here's a mockup where you can see TabWidget displaying both artwork and text tabs:

![untitled](https://cloud.githubusercontent.com/assets/8179031/13169359/24f4f8f4-d6e5-11e5-8396-ea5d28dbb64e.png)

I'm up for improving them a little, but I guess I won't be able to do anything approching what @RebeccaDeField could do.

**Comment by tresf** [comment]
@StCyr thanks.  I agree they could use some touch-up, but I don't think they harm the interface.  Making them bit wider would help. :+1:

**Comment by claell** [comment]
Maybe use auto width, so each tab is sized that it fills the whole space.

**Comment by StCyr** [comment]
@claell tabs are not widgets (only the tab bar is), so it's not that easy to have auto-width. Basically, I  can't use any Qt helper, I would have to handle all that stuff myself within the tab bar.

@tresf we could make them 36 pixels wide. That would make place for 7 such tabs, and thus one seat would be free. But should I do it or would @RebeccaDeField do it? Personally, I don't see how I could make them nicer with only 14 pixels heights :-s

**Comment by tresf** [comment]
> @tresf we could make them 36 pixels wide. That would make place for 7 such tabs, and thus one seat would be free. But should I do it or would @RebeccaDeField do it?

Depends.... Pixmap would be Rebecca, C++ would be you.

> Personally, I don't see how I could make them nicer with only 14 pixels heights :-s

I struggle with the pixel-art too... Perhaps Rebecca can help.

**Comment by claell** [comment]
Is it bad to increase the height a bit? Might look better then.

I have absolutely no experience with Qt, but one cannot create tabs with auto width (and corresponding icon centred) with it?

**Comment by tresf** [comment]
> Is it bad to increase the height a bit? Might look better then.

Not if it is mandatory, no a few pixels wouldn't harm anything, but we'd have to make sure we extensively test how it impacts the plugins. :+1: 

> I have absolutely no experience with Qt, but one cannot create tabs with auto width (and corresponding icon centred) with it?

I've never coded the tabs specifically, but I can confirm that the layout engine is much less robust when compared to other GUI libraries I've used.  Link for reference... https://github.com/LMMS/lmms/issues/387#issuecomment-36679441

**Comment by claell** [comment]
Alright, so this is probably not possible to achieve with Qt?

**Comment by tresf** [comment]
> Alright, so this is probably not possible to achieve with Qt?

Anything is possible, it's complexity just puts it out of scope. :+1:

**Comment by claell** [comment]
Ok

**Comment by midi-pascal** [comment]
@StCyr Did you try <code>QTabBar::setExpanding(true);</code> ?

**Comment by midi-pascal** [comment]
@StCyr Ok I found how you can fix the tabs width:
In TabWidget::addTab() you can fix the tab width by calling 
<code>	widgetDesc d = { _w, _name, 30} ;</code>
instead of 
<code> 	widgetDesc d = { _w, _name, fontMetrics().width( _name ) + 10 } ;</code>
I put 30 as a value for demo purpose only.

**Comment by RebeccaDeField** [comment]
If the tabs are 14px high, should the icons be 12px to give some space around the icons? I have made 14px icons before, but 12px might be a bit small. I can still give it a try though.

**Comment by StCyr** [comment]
Hi @RebeccaDeField 

thanks for your answer. I'll have a look how I can make this tabs taller.

Cyrille

**Comment by StCyr** [comment]
Hi,

Here's an update of the requested features for these artwork tabs:

1) Making them taller;
2) Making them wider;
3) Tooltip

===================================

**1) Making them taller:**

I've looked a little at how to make the tabs taller. But, it's not as easy as it looks.

The problem is that this TabWidget thing is used in various places, and changing its height involves changing the height of all the involved parent widgets.

I think it might be easier to just handle both heights: Either you draw a "legacy" text tabbar, and its height is 14 pixels, either you draw a new graphical tabbar, and you get, for example, a height of 20 pixels. 

This new graphical tabbar would be used only here, so that for the moment only changing this parent's height (InstrumentTrackWindow) would be needed.

**2) Making them wider:**

I've a first PoC of an "autowidth" function. This requires PNG pixmaps with alpha channel.

**2) Tooltip:**

I haven't progressed on this front.

**Comment by claell** [comment]
Sounds good so far, especially "autowidth" and hopefully the heigth thing can get adjusted as well (if needed, maybe 14px is good as well).

**Comment by RebeccaDeField** [comment]
I will still give the 12px icons and shot and let you know how it works.

**Comment by StCyr** [comment]
Ok @RebeccaDeField great! Please use transparent background so that I can easily place the icons evently in the tabbar.

In the mean time, I've implemented a 20 pixels tall tabbar. Here are some mockups (only the FX, MIDI, and MISC tabs have 20 pixels tall artworks, but it gives already an idea):

![mockup_20pxl](https://cloud.githubusercontent.com/assets/8179031/13205324/29ef2368-d8e6-11e5-800b-2bcc67474fb9.png)
![mockup3_20pxl](https://cloud.githubusercontent.com/assets/8179031/13205323/29eea58c-d8e6-11e5-98ff-c5b17f37843a.png)
![mockup2_20pxl](https://cloud.githubusercontent.com/assets/8179031/13205322/29edfb78-d8e6-11e5-89ba-a24a1c4f349d.png)

PoC will be updated in a few minutes: https://github.com/StCyr/lmms/tree/StCyr_ArtworkTabs

**Comment by RebeccaDeField** [comment]
![tabicons](https://cloud.githubusercontent.com/assets/7960169/13206616/d2c30bec-d8b6-11e5-911e-d0e5eca347df.png)
12x12 pixel perfect transparent vector icons as requested :)

**Comment by StCyr** [comment]
@RebeccaDeField they are wonderful. Where can I grab them? The link you provided in your previous comment doesn't work for me.

Cyr

**Comment by RebeccaDeField** [comment]
@StCyr  Sorry about that. Should work now.

**Comment by StCyr** [comment]
Hi @RebeccaDeField 

I'm afraid it still doesn't work.

It may be that I don't know the correct method to retrieve these files, but I think you haven't shared them.

The following page explains one way to do that: https://www.dropbox.com/help/167

Cyr

**Comment by claell** [comment]
@StCyr I managed to download the file with this link.

**Comment by StCyr** [comment]
ok got it :+1:

**Comment by StCyr** [comment]
Hi,

Here's a screenshot with @RebeccaDeField 's icons.

Tooltip (in green) has also been implemented.

![untitled](https://cloud.githubusercontent.com/assets/8179031/13216926/43da8a14-d95f-11e5-9818-50932623fd47.png)

You can compile it from https://github.com/StCyr/lmms/tree/StCyr_ArtworkTabs

Please give me your feedback.

Cyrille

**Comment by claell** [comment]
Already looks pretty good. Things I noticed:
- The selection background seems to overlap the tabbar in height.
- The icon on the right looks like it is not really centered, but a bit too much on the right. Is this an issue with the auto-width prototype?

**Comment by tresf** [comment]
> Tooltip (in green) has also been implemented.
> ![untitled](https://cloud.githubusercontent.com/assets/8179031/13216926/43da8a14-d95f-11e5-9818-50932623fd47.png)

That's what I call teamwork.  Agreed with all points made, but unless someone disagrees, marking for the 1.2 milestone.  This is coming together nicely.

**Comment by Umcaruje** [comment]
I'll check @StCyr's branch and play around with it

**Comment by StCyr** [comment]
@Umcaruje please do. Shall I let you manage merging? Or shall I send a PR?

@claell you're right. I've fine tuned the position a little bit. Here's a new screenshot:

![untitled](https://cloud.githubusercontent.com/assets/8179031/13220012/db6ba372-d973-11e5-80f5-920aabc93931.png)

Cyrille

**Comment by Umcaruje** [comment]
>@Umcaruje please do. Shall I let you manage merging? Or shall I send a PR?

You should send a PR, I'll check out the branch to test it all out.

**Comment by claell** [comment]
@StCyr Looks fine now.

**Comment by zonkmachine** [comment]
I think this looks really good and I agree on the 1.2 milestone. The only tiny issue I have is with the second icon, the sine wave, which makes me think the tab is for sound generation. I think it would be better to have a small envelope graph there or something else related to sound modulation.

**Comment by tresf** [comment]
> I think it would be better to have a small envelope graph there or something else related to sound modulation

Good point. :+1:

**Comment by DeRobyJ** [comment]
I suggest an icon that resambles an envelope as theories book show it
http://libremusicproduction.com/sites/default/files/answers/adsrenvelope.png

Our envelope graph doesn't have that long line in sustain, because amplitude it's intended to be still until the key is released, but that will remind more people about an adsr than the one here does: https://cloud.githubusercontent.com/assets/8918426/12841345/7785bcfc-cbeb-11e5-8391-81c1bdb36587.png

**Comment by RebeccaDeField** [comment]
I tried to fit the entire asdr into the icon, but it wasn't recognizable in the space I have available. I tried taking out the release. Please see the pic bellow and let me know if this works.

![tabicons](https://cloud.githubusercontent.com/assets/7960169/13243049/e738246e-d9af-11e5-9ead-4a8e579ab834.png)

**Comment by tresf** [comment]
@RebeccaDeField the only criticism is the [attack seems too low](https://cloud.githubusercontent.com/assets/6345473/13243267/31b187c2-d9cb-11e5-8524-919e86c4f77d.png) as envelope is normally drawn on a flat line.  Perhaps even [our default envelope shape](https://cloud.githubusercontent.com/assets/6345473/13243280/57bb2aa4-d9cb-11e5-85c0-495d6761e510.png) is sufficient if you have enough room for it...

**Comment by zonkmachine** [comment]
I would prefer a more classic envelope graph. Something like this.
![zenvelope](https://cloud.githubusercontent.com/assets/6368949/13243670/00f8aa3a-da01-11e5-9374-e1a51cf87edc.png)
OK. It is hard to draw at those sizes...

**Comment by StCyr** [comment]
@zonkmachine +1 for your env/lfo icon

**Comment by RebeccaDeField** [comment]
@zonkmachine I'm not a producer myself (I don't know much about envelopes and such), so your idea is very helpful. I played with that concept a bit and this is what I came up with.
![tabicons](https://cloud.githubusercontent.com/assets/7960169/13245535/d52bc0d8-d9c2-11e5-9773-33966e7adaed.png)

**Comment by DeRobyJ** [comment]
OK that's something that surely reminds about adsr.
The line is a bit tighter than the other icons, though... What to do...
I'm OK with this one

**Comment by DeRobyJ** [comment]
Oh sorry just saw zonkmachine's one
Well, that's slightly better :)
(Sorry Rebecca D:)

**Comment by RebeccaDeField** [comment]
Here is another option. I think you guys can decide which one makes more sense to you.
![tabicons1](https://cloud.githubusercontent.com/assets/7960169/13246039/fb16d64a-d9c5-11e5-98db-9d80e8f7a211.png)

![tabicons](https://cloud.githubusercontent.com/assets/7960169/13246043/0642e0ae-d9c6-11e5-8b01-ada9e23305a3.png)

I put both here for comparison.

**Comment by BaraMGB** [comment]
The thicker lined fits better in row.

**Comment by StCyr** [comment]
I also prefer the thicker line

**Comment by midi-pascal** [comment]
:+1: for the thicker one.

**Comment by DeRobyJ** [comment]
:+1: thicker one

**Comment by tresf** [comment]
:+1: thick

**Comment by zonkmachine** [comment]
:+1: thick

**Comment by RebeccaDeField** [comment]
Well, I guess that has been decided then :) Here is the download link --> https://www.dropbox.com/s/trd1ombn4kws03e/TabIcons.svg?dl=0

**Comment by tresf** [comment]
This is waiting on a few items... 
1. Update artwork at #2599
1. Fix Travis build issues with our Qt5 builds

Once these items are completed, we'll be happy to merge and close this issue out.

**Comment by StakeoutPunch** [comment]
I believe it would make more sense if the plugin (instrument tab) icon was used for the FX _plugin_ tab, and using the icon originally meant for the envelope tab for the instrument control tab. My reasoning is that the instrument tab controls the actual generation of the sound (sound wave/sine), while the FX tab adds plugins (puzzle piece) onto that source audio.

Mockup:
![tco_mock_1](https://cloud.githubusercontent.com/assets/6579031/13303380/ad7b15cc-db14-11e5-844a-23b5f7199712.png)

**Comment by claell** [comment]
The white highlighting behind a selected tab still overlaps the tabbar about one pixel in height. Is this wanted?

**Comment by StCyr** [comment]
@claell no i isn't wanted. Where do you think it overlaps? At the top of the tabbar, that's it?

**Comment by claell** [comment]
If I zoom it to the highlighting behind the envelope icon the light grey space seems to overlap about 1px at the top and about 2px at the bottom.

**Comment by StCyr** [comment]
Hi @claell 

You're right. 

I've fine tuned the positioning a little bit.

If we want to be picky there's still an issue. But, how hard is it to be solved?

Here's the latest version:

![mockup4](https://cloud.githubusercontent.com/assets/8179031/13333642/4256960a-dc08-11e5-8d52-85fe41adbfa5.png)

@tresf I've fixed the QT5 compilation issue (and updated the artwork). Now, there might still be some fine tuning left.

**Comment by tresf** [comment]
> I've fine tuned the positioning a little bit.

If no one is attached to that embossed background image, let's see what we can do with the larger tabs.  Feel free to steal a pixel or two from the VOL/PAN/PITCH area. :+1:   I'd even go as far as to ask @Umcaruje or @RebeccaDeField to assist here.

> @tresf I've fixed the QT5 compilation issue

:+1:

**Comment by Umcaruje** [comment]
I'll be glad to asist. I vote we get rid of the gradient. Gradients on such small surfaces take the clarity off.

**Comment by claell** [comment]
@StCyr Now it looks like the whole tab got smaller? At least some icons are overlapping it. But maybe this was already the case before the cange. I am also for increasing the tab size.

And @Umcaruje if you think that gradients are not good on small surfaces, might it make sense to remove them from all those surfaces? Would need some rework of design, maybe one should give lmms a complete overhaul designwise.

**Comment by StCyr** [comment]
@claell I didn't change the tab's height yet; It was already the case.

Hmm, the idea of changing the tab 'size was actually a little bit invasive 'cause TabWidgets are used at some other places too. For the record, here's the list of files were they may be in use:

cyr@opennms-dev:~/lmms$ grep -Rl TabWidget src/
src/gui/widgets/InstrumentSoundShapingView.cpp
src/gui/widgets/TabWidget.cpp
src/gui/AudioDeviceSetupWidget.cpp
src/gui/ControllerConnectionDialog.cpp
src/gui/SetupDialog.cpp
src/gui/CMakeLists.txt
src/gui/MidiSetupWidget.cpp
src/gui/dialogs/about_dialog.ui
src/core/NotePlayHandle.cpp
src/tracks/InstrumentTrack.cpp

But, with my changes artwork tabs can be treated differently than text tabs. And, artworks tabs would currently only be used in "src/tracks/InstrumentTrack.cpp" (concern of this issue).

So, I guess I can change only this tab's height and also only remove gradient from this tab.

There was also this idea of @StakeoutPunch to change the icon for the plugin and fx tabs. I'm not in favor of changing the fx tab's icon as it has the same look as the "show/hide FX mixer" icon in LMMS' general toolbar. But, its reasoning for the plugin tab (the plugin tab is the sound source for the InstrumentTrack) actually makes sense. I'm not convinced though that the simple sine-wave icon is good enough. Earlier @Spekular also suggested an icon like the followings for this tab:

![untitled](https://camo.githubusercontent.com/927e46ba0b25734e8461f530dcebaefd11649f54/687474703a2f2f7777772e736f6c61727374696b2e636f6d2f73697465732f64656661756c742f66696c65732f636f6e74656e742f77617465725f67656e657261746f725f69636f6e312e706e67)

![untitled](https://camo.githubusercontent.com/e0f1e12ddce83b70ab07b4e7eed4adef39681858/687474703a2f2f61332e6d7a7374617469632e636f6d2f6e7a2f7233302f507572706c65312f76342f35392f38342f36652f35393834366564612d313564342d306639312d626638642d6465386336636238326636332f69636f6e313735783137352e6a706567)

Looking for "turbine" icons on google also returns interesting results: https://www.google.com/search?q=turbine&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiknt7ciZXLAhXH1xQKHTr6B-sQ_AUIBygB&biw=1680&bih=953#q=turbine&tbm=isch&tbs=isz:i

**Comment by StCyr** [comment]
I've updated my branch.

TabWidgets with artwork tabs are now 17 pixels tall while TabWidgets with text tabs are kept to 14 pixels height.

**Comment by Umcaruje** [comment]
>And @Umcaruje if you think that gradients are not good on small surfaces, might it make sense to remove them from all those surfaces? Would need some rework of design, maybe one should give lmms a complete overhaul designwise.

That's what we're doing in #2587 ;) And yes, I'll look into flattening all of those line things.

**Comment by claell** [comment]
@Umcaruje thanks for the link, did not know about this.

**Comment by tresf** [comment]
I think this was very close to be finished and the PR over at https://github.com/LMMS/lmms/pull/2599 has some commits since the last discussion.  @StCyr can we get an update on this?

**Comment by StCyr** [comment]
Hello @tresf 

Your last request (25 Feb. in this issue) was to investigate the possibilty of having a larger (ie: taller) TabWidget for this specific toolbar.

I made this in commits dccf9f411996c8c866ff1086b65f15020ef08cd9 and https://github.com/LMMS/lmms/pull/2599/commits/7aad86efe0ff4e17e8cb5d3c7da7f823b8e5a065

In addition:

- I've reverted the change I made in commit 64efd68bfdcda7e12a6c2abaf222badcb9f1c0ad to fix issue https://github.com/LMMS/lmms/issues/2519 (commit https://github.com/LMMS/lmms/pull/2599/commits/fc157554c189846b89725ad06508ae6720c1a22a)
- I've added a tooltip functionality so that one can have additional meaningful text when hovering over a tab (commit https://github.com/LMMS/lmms/pull/2599/commits/3f094407491897f1e79f99bdd5211179b675967f)
-  I've changed the ENV/LFO tab's artwork to a more standard ADSR graph as it had more proponents (see zonkmachine's comment of Feb. 22 in this issue for the discussion starting point on this particular topic). That's the purpose of commit https://github.com/LMMS/lmms/pull/2599/commits/857b6ed93f0f526ef82a91871bfcbb80b7bb4b89
- I've fixed a bug when a TabWidget with only one tab would generate a SIGFPE error (division by zero). That's the purpose of commit https://github.com/LMMS/lmms/pull/2599/commits/d9edef324f62b27b58756f9a2f845928896ac8ba

By re-reading the discussion, it seems that 2 other requests might have been raised:

- Flatening the TabWidget / Remove the gradients: The highlighted tabs are flatened, though I didn't changed anything related to the non-highlighted tabs' gradient.
- The artwork of the first tab (PLUGIN tab / jigsaw puzzle artwork) received some criticisms but no formal request.

Cheers,

Cyrille

**Comment by tresf** [comment]
@StCyr thanks for the detailed response. I'd like to get this merged in soon.  Can you be so kind as to provide a final mockup so that we can discuss please?

@Umcaruje can you please help review this (tagging you since you've been administering the theme changes).

**Comment by Umcaruje** [comment]
Reviewed the code, it looks ok to me, I made some remarks over magic numbers and the fact that the tab widget is not themeable using css, but one could argue that the themeablity is out of scope for that PR, but it would definitely be a nice thing to have.

I haven't tested the branch yet though, I'll try to get around it, but can't make any promises :)

**Comment by Umcaruje** [comment]
Oh and I just read that gradients are still there, I'll look into that too when I get around testing this.

**Comment by StCyr** [comment]
Hello @Umcaruje, @Tresf,

As I said, the highlighting of tabs doesn't have gradient. I'll have a look
how to remove all TabWidget gradients.

For what concerne the themeability, I don't have any idea how to do that.
If you have any documentation pointers, and this isn't too much work ;-),
I'll be glad to do it.

I'll look at your other remarks and address them eventually.

Cheers,

Cyrille

2016-04-26 23:32 GMT+02:00 Umcaruje <notifications@github.com>:

**Comment by tresf** [comment]
> For what concerne the themeability, I don't have any idea how to do that.
> If you have any documentation pointers, and this isn't too much work ;-)

Here's a good recent example of how @Umcaruje's been doing it... https://github.com/LMMS/lmms/commit/f787982d9ab80fce38440076b0e81ce44eecf1dc

**Comment by Umcaruje** [comment]
Closed via #3569
